File: completions-streaming/final/typescript/src/browser/renderer.ts
/**
 * CLI typing-effect renderer for streaming completions.
 *
 * Writes each token to stdout without newlines as it arrives, creating
 * a real-time typing effect in the terminal (Node.js).
 */

import { makeStreamingRequest, parseSseLine, extractTokenContent } from "../streaming/streamClient.js";

/**
 * Stream a completion and print tokens to stdout with a typing effect.
 *
 * @param message - The user message to send
 * @param token - Valid Bearer token for authorization
 */
export async function renderStreamToTerminal(message: string, token: string): Promise<void> {
  process.stdout.write(`Prompt: ${message}\n\nResponse: `);

  const response = await makeStreamingRequest(message, token);
  const reader = response.body!.getReader();
  const decoder = new TextDecoder();

  let buffer = "";
  let totalTokens = 0;
  let finishReason = "unknown";

  try {
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split("\n");
      buffer = lines.pop() ?? "";

      for (const line of lines) {
        const chunk = parseSseLine(line);
        if (chunk === null) continue;
        if (chunk === "[DONE]") break;

        const content = extractTokenContent(chunk);
        if (content) {
          process.stdout.write(content);
          totalTokens += 1;
        }

        const choices = chunk?.choices;
        if (choices && choices[0]?.finish_reason) {
          finishReason = choices[0].finish_reason;
        }
      }
    }
  } finally {
    reader.releaseLock();
  }

  process.stdout.write(`\n\n[${totalTokens} tokens, finish_reason=${finishReason}]\n`);
}

File: completions-streaming/final/go/pkg/browser/renderer.go
// Package browser provides a CLI typing-effect renderer for streaming completions.
package browser

import (
	"bufio"
	"fmt"
	"os"

	"completions-streaming/pkg/streaming"
)

// RenderStreamToTerminal streams a completion and prints tokens to stdout
// without newlines, creating a typing effect in the terminal.
//
// Each token is written immediately via os.Stdout so the user sees text
// appear token-by-token without waiting for the full response.
func RenderStreamToTerminal(message, token string) error {
	fmt.Printf("Prompt: %s\n\nResponse: ", message)

	resp, err := streaming.MakeStreamingRequest(message, token)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	totalTokens := 0
	finishReason := "unknown"

	scanner := bufio.NewScanner(resp.Body)
	for scanner.Scan() {
		line := scanner.Text()
		parsed := streaming.ParseSSELine(line)
		if parsed == nil {
			continue
		}
		if s, ok := parsed.(string); ok && s == "[DONE]" {
			break
		}
		chunk, ok := parsed.(*streaming.SSEChunk)
		if !ok {
			continue
		}

		content := streaming.ExtractTokenContent(chunk)
		if content != "" {
			fmt.Fprint(os.Stdout, content)
			totalTokens++
		}

		if len(chunk.Choices) > 0 && chunk.Choices[0].FinishReason != nil {
			finishReason = *chunk.Choices[0].FinishReason
		}
	}

	if err := scanner.Err(); err != nil {
		return fmt.Errorf("error reading stream: %w", err)
	}

	fmt.Printf("\n\n[%d tokens, finish_reason=%s]\n", totalTokens, finishReason)
	return nil
}
